### PR TITLE
Allow empty excludedReferrers on the global setting

### LINKED ISSUE #19833: Allow empty excludedReferrers on the global setting
reported in https://forum.matomo.org/t/the-address-is-not-a-valid-url/47725 and reproduced on 4.12.0

If e.g. the "Global list of excluded referrers" is empty, the form can't be saved:
grafik

related to #19302

### PR DESCRIPTION
**Pull request by peterhashair**
### Description:

Fixes: #19833 

allow empty excludedReferrers on the global setting

Also, I notice when the setting changed successfully, there is no notification popup, should we add one?

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/SitesManager/API.php [hunk 1]
```diff
@@ -1155,23 +1155,25 @@ public function setGlobalExcludedReferrers(string $excludedReferrers): void
...
-        // update option
-        Option::set(self::OPTION_EXCLUDED_REFERRERS_GLOBAL, $excludedUrls);
+            // update option
+            Option::set(self::OPTION_EXCLUDED_REFERRERS_GLOBAL, $excludedUrls);
 
-        // make sure tracker cache will reflect change
-        Cache::deleteTrackerCache();
+            // make sure tracker cache will reflect change
+            Cache::deleteTrackerCache();
+        }
```
Comment (sgiehl) [inline plugins/SitesManager/API.php]: That solution actually can't work. It means that the option is actually only updated if a value was provided. So storing an empty value is only skipped and the old value remains.

File: plugins/SitesManager/API.php [hunk 2]
```diff
@@ -1155,23 +1155,25 @@ public function setGlobalExcludedReferrers(string $excludedReferrers): void
 
         $excludedUrls = $this->checkAndReturnCommaSeparatedStringList($excludedReferrers);
 
-        foreach (explode(',', $excludedUrls) ?: [] as $url) {
```
Comment (sgiehl) [inline plugins/SitesManager/API.php:1158]: I guess the `?: []` check was already meant to catch empty values, but was implemented incorrectly.
Simply using this here, should already fix everything:
```
foreach (!empty($excludedUrls) ? explode(',', $excludedUrls) : [] as $url) {
```

**Review by sgiehl — CHANGES_REQUESTED** [review]
Might maybe be worth adding a test case for storing an empty value, to ensure the solution really works.

**Comment by peterhashair** [comment]
@sgiehl that makes more sense, add a couple of tests, just wondering `example a` should isLooksLikeURl accept that?

File: plugins/SitesManager/API.php [hunk 3]
```diff
@@ -1155,23 +1155,24 @@ public function setGlobalExcludedReferrers(string $excludedReferrers): void
...
-                throw new Exception(Piwik::translate('SitesManager_ExceptionInvalidUrl', [$url]));
+            foreach (!empty($excludedUrls) ? explode(',', $excludedUrls) : [] as $url) {
+                // We allow urls to be provided:
+                // - fully qualified like http://example.url/path
+                // - without protocol like example.url/path
+                // - with subdomain wildcard like .example.url/path
+                $prefixedUrl = 'https://'.ltrim(preg_replace('/^https?:\/\//', '', $url), '.');
+                $parsedUrl = @parse_url($prefixedUrl);
+                if (false === $parsedUrl || filter_var($prefixedUrl,
+                        FILTER_VALIDATE_URL) === false || !UrlHelper::isLookLikeUrl($prefixedUrl)) {
```
Comment (sgiehl) [inline plugins/SitesManager/API.php]: Are you aware that `FILTER_VALIDATE_URL` only allows ascii domains and might fail for internationalized domains?

**Review by sgiehl — CHANGES_REQUESTED** [review]
@peterhashair `example a` might not be a valid record but we need to take care to be not too restrictive, as this might produce new problems.

**Review by sgiehl — APPROVED** [review]
I've fixed the code indent, that was wrong after your changes. Looks good now.
